lms/djangoapps/experiments/flags.py

```diff
@@ -0,0 +1,133 @@
+
+    But you'll also have smaller related flags that can force bucketing certain users into
+    specific buckets of your experiment. Those can be set using a waffle named like
+    "main_flag.BUCKET_NUM" (e.g. "course_experience.animated_exy.0") to force
+    users that pass the first main waffle check into a specific bucket experience.
+
+    If you pass this flag a course key, tracking calls to segment will be made per-course-run
+    (rather than one call overall) and will include the course key.
+
+    You can also control whether the experiment only affects future enrollments by setting
+    an ExperimentKeyValue model object with a key of 'enrollment_start' to the date of the
+    first enrollments that should be bucketed.
+
+    Bucket 0 is assumed to be the control bucket.
+
+    .. no_pii:
+    """
+    def __init__(self, waffle_namespace, flag_name, num_buckets=2, experiment_id=None, **kwargs):
+        super().__init__(waffle_namespace, flag_name, **kwargs)
+        self.num_buckets = num_buckets
+        self.experiment_id = experiment_id
+        self.bucket_flags = [
+            CourseWaffleFlag(waffle_namespace, '{}.{}'.format(flag_name, bucket), flag_undefined_default=False)
+            for bucket in range(num_buckets)
+        ]
+
+    def _cache_bucket(self, key, value):
+        request_cache = RequestCache('experiments')
+        request_cache.set(key, value)
+        return value
+
+    def get_bucket(self, course_key=None, track=True):
+        """
+        Return which bucket number the specified user is in.
+
+        Bucket 0 is assumed to be the control bucket and will be returned if the experiment is not enabled for
+        this user and course.
+        """
+        # Keep some imports in here, because this class is commonly used at a module level, and we want to avoid
+        # circular imports for any models.
+        from experiments.models import ExperimentKeyValue
+        from student.models import CourseEnrollment
+
+        request = get_current_request()
+        if not request:
+            return 0
+
+        # Use course key in experiment name to separate caches and segment calls per-course-run
+        experiment_name = self.namespaced_flag_name + ('.{}'.format(course_key) if course_key else '')
+
+        # Check if we have a cache for this request already
+        request_cache = RequestCache('experiments')
+        cache_response = request_cache.get_cached_response(experiment_name)
+        if cache_response.is_found:
+            return cache_response.value
+
+        # Check if the main flag is even enabled for this user and course.
+        if not self._is_enabled(course_key):  # grabs user from the current request, if any
+            return self._cache_bucket(experiment_name, 0)
+
+        # Check if the enrollment should even be considered (if it started before the experiment wants, we ignore)
+        if course_key and self.experiment_id is not None:
```

Comment:
I was just discussing how we want the experiment to roll out with @jmyatt and Kaitlin, and we realized that it would be useful to also have a date after which all users are bucketed into treatment. That will allow us to follow the holdback group for a longer period while still at some point exposing the feature to all new enrollments.

I don't think that necessarily needs to go in this PR, but it could.

Reply:
I think I overheard snippets of that conversation -- my vote would be to save that for a followon PR, just because this PR is blocking us testing the features. It should be a simple addon.

Reply:
Yup, I agree.